text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+    public static UnionType FromFoo(Foo value) { /* ... */ }
+    public static UnionType FromMapOfBar(IReadOnlyDictionary<Bar> value) { /* ... */ }
+    public static UnionType FromListOfBaz(IReadOnlyList<Baz> value) { /* ... */ }
+
+    public Foo? Foo { get; }
+    public IReadOnlyDictionary<Bar>? MapOfBar { get; }
+    public IReadOnlyList<Baz>? ListOfBaz { get; }
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public final class UnionType {
+    public static UnionType fromFoo(Foo value) { /* ... */ }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) { /* ... */ }
+    public static UnionType fromListOfBaz(Baz... value) { /* ... */ }
+    public static UnionType fromListOfBaz(List<Baz> value) { /* ... */ }
+
+    public @Nullable Foo asFoo() { /* ... */ }
+    public @Nullable Map<String, Bar> asMapOfBar() { /* ... */ }
+    public @Nullable List<Baz> asListOfBaz() { /* ... */ }
+  }
+  ```
+
+The _jsii Kernel_ will need adjustments in order to better support those types. The generated libraries will provide a
+new argument to the _jsii kernel_'s `load` API if it's code was generated with `--union-like-types` enabled. This way,
+the kernel will be bale to reference the declared _type union_'s information when annotating instances that are passed
+from **Javascript** to the host application, removing all ambiguity regarding which type is serialized.
+
+# Drawbacks
+
+This approach might be expensive to implement as it might require fundamentally changing how the `jsii` compiler works.
+It currently operates on the `tsc` _Type Checker_ API, which preprocesses type aliases away in order to have a more
+streamlined type system to reason over. In order to property process aliased _type unions_, it may be necessary to fall
+back on the `Node` API, which allows more direct inspection of the original source code, however provides fewer
+inspection capabilities (since the type system has not been fully resolved at the time a `Node` is processed).
+
+# Rationale and Alternatives
+
+Improving support for _type unions_ appears to be the best way going forward, as complete removal of type unions has
+proven to be challenging from an API design pespective: it tends to make the experience significantly worse for dynamic
+languages (as boilerplate-heavy patterns must be leveraged as a replacement for _type unions_), and would reduce the
+flexibility in language-specific code generation for _union-like_ types (as those would be bound to the **TypeScript**
+API). Full support of _type unions_ as a type kind in the _jsii type model_ achieves better outcomes:
+
+- it ensures consistency around _union-like_ types
+- it does not degrade the developer experience in languages with _native type_ union support
+- it requires significantly less effort from library authors
+
+This feature could however not be introduced as part of CDK v1, as the change of static (and dynamic) type it requires
```

Comment:
As mentioned above, the only thing that can't be introduce on 1.x is changing the types of .NET and Java from Object to the union-like. But we can technically already generate and accept the union like on v1.x, no?

Another idea is to only generate union-like classes when the typescript code uses a union type alias. Then, change our L1 code generators to generate the union type aliases only for new properties introduced after some CFN spec version (not sure how to do that). 

The benefit is that we can basically start migrating towards this (and baking it!) without breaking anyone. Then, when we are ready, in the next CDK major version we can flip the switch and generate union type aliases for everything.

Reply:
> Another idea is to only generate union-like classes when the typescript code uses a union type alias. Then, change our L1 code generators to generate the union type aliases only for new properties introduced after some CFN spec version (not sure how to do that).

So technically, we can have `jsii-diff` treat converting from an "inline union" to an aliased one as a breaking change. We can then introduce an `awslint` rule that requires aliased unions (and add exceptions for all existing inline unions) so we stop proliferation.

It has the benefit of allowing out-of-band (meaning directly on v1) breaking changes on those APIs that are returning struct/interface unions, and are **known** to currently be unusable in **C#** or **Java** right now (since their broken nature would mean we cannot be breaking any usage - or at least not more than it already was).